File: autara-lib/src/pda.rs
use arch_program::pubkey::Pubkey;

#[inline(always)]
pub fn market_seed_without_bump<'a>(
    curator: &'a Pubkey,
    supply_mint: &'a Pubkey,
    collateral_mint: &'a Pubkey,
    index: &'a [u8; 1],
) -> [&'a [u8]; 5] {
    [
        b"market",
        curator.as_ref(),
        supply_mint.as_ref(),
        collateral_mint.as_ref(),
        index,
    ]
}

#[inline(always)]
pub fn market_seed_with_bump<'a>(
    curator: &'a Pubkey,
    supply_mint: &'a Pubkey,
    collateral_mint: &'a Pubkey,
    index: &'a [u8; 1],
    bump: &'a [u8; 1],
) -> [&'a [u8]; 6] {
    [
        b"market",
        curator.as_ref(),
        supply_mint.as_ref(),
        collateral_mint.as_ref(),
        index,
        bump,
    ]
}

#[inline(always)]
pub fn supply_position_seed<'a>(market: &'a Pubkey, authority: &'a Pubkey) -> [&'a [u8]; 3] {
    [b"supply_position", market.as_ref(), authority.as_ref()]
}

#[inline(always)]
pub fn supply_position_seed_with_bump<'a>(
    market: &'a Pubkey,
    authority: &'a Pubkey,
    bump: &'a [u8; 1],
) -> [&'a [u8]; 4] {
    [
        b"supply_position",
        market.as_ref(),
        authority.as_ref(),
        bump,
    ]
}

#[inline(always)]
pub fn borrow_position_seed<'a>(market: &'a Pubkey, authority: &'a Pubkey) -> [&'a [u8]; 3] {
    [b"borrow_position", market.as_ref(), authority.as_ref()]
}

#[inline(always)]
pub fn borrow_position_seed_with_bump<'a>(
    market: &'a Pubkey,
    authority: &'a Pubkey,
    bump: &'a [u8; 1],
) -> [&'a [u8]; 4] {
    [
        b"borrow_position",
        market.as_ref(),
        authority.as_ref(),
        bump,
    ]
}

#[inline(always)]
pub fn find_supply_position_pda(
    program_id: &Pubkey,
    market: &Pubkey,
    authority: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(&supply_position_seed(market, authority), program_id)
}

#[inline(always)]
pub fn find_borrow_position_pda(
    program_id: &Pubkey,
    market: &Pubkey,
    authority: &Pubkey,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(&borrow_position_seed(market, authority), program_id)
}

#[inline(always)]
pub fn find_market_pda(
    program_id: &Pubkey,
    curator: &Pubkey,
    supply_mint: &Pubkey,
    collateral_mint: &Pubkey,
    index: u8,
) -> (Pubkey, u8) {
    Pubkey::find_program_address(
        &market_seed_without_bump(curator, supply_mint, collateral_mint, &[index]),
        program_id,
    )
}

#[inline(always)]
pub fn global_config_seed() -> [&'static [u8]; 1] {
    [b"global_config"]
}

#[inline(always)]
pub fn global_config_seed_with_bump(bump: &[u8; 1]) -> [&[u8]; 2] {
    [b"global_config", bump]
}

#[inline(always)]
pub fn find_global_config_pda(program_id: &Pubkey) -> (Pubkey, u8) {
    Pubkey::find_program_address(&global_config_seed(), program_id)
}

File: src/lib/autara/gateway/arch/pda.ts
/**
 * PDA derivation — direct port of autara-lib/src/pda.rs.
 *
 * Seed layout must match the Rust program byte-for-byte:
 *   market:          ["market", curator, supply_mint, collateral_mint, index]
 *   supply_position: ["supply_position", market, authority]
 *   borrow_position: ["borrow_position", market, authority]
 *   global_config:   ["global_config"]
 */

import type { PublicKeyStr } from "../../domain/types/common";

// These constants match the Rust seed literals exactly
const MARKET_SEED = new TextEncoder().encode("market");
const SUPPLY_POSITION_SEED = new TextEncoder().encode("supply_position");
const BORROW_POSITION_SEED = new TextEncoder().encode("borrow_position");
const GLOBAL_CONFIG_SEED = new TextEncoder().encode("global_config");

/**
 * Generic PDA finder — wraps the platform-specific findProgramAddress.
 * Accepts a function that does the actual cryptographic derivation
 * so this module stays dependency-free.
 */
export type FindProgramAddressFn = (
  seeds: Uint8Array[],
  programId: Uint8Array,
) => [Uint8Array, number];

export interface PdaDeriver {
  findMarketPda(
    curator: PublicKeyStr,
    supplyMint: PublicKeyStr,
    collateralMint: PublicKeyStr,
    index: number,
  ): [PublicKeyStr, number];

  findSupplyPositionPda(
    market: PublicKeyStr,
    authority: PublicKeyStr,
  ): [PublicKeyStr, number];

  findBorrowPositionPda(
    market: PublicKeyStr,
    authority: PublicKeyStr,
  ): [PublicKeyStr, number];

  findGlobalConfigPda(): [PublicKeyStr, number];
}

/**
 * Create a PDA deriver for a given program ID.
 *
 * @param programId - The Autara program's public key as bytes
 * @param findProgramAddress - Platform-specific PDA derivation (e.g. from @solana/web3.js or arch-sdk)
 * @param encodeKey - Convert PublicKeyStr to bytes
 * @param decodeKey - Convert bytes back to PublicKeyStr
 */
export function createPdaDeriver(
  programId: Uint8Array,
  findProgramAddress: FindProgramAddressFn,
  encodeKey: (key: PublicKeyStr) => Uint8Array,
  decodeKey: (bytes: Uint8Array) => PublicKeyStr,
): PdaDeriver {
  return {
    findMarketPda(curator, supplyMint, collateralMint, index) {
      const [address, bump] = findProgramAddress(
        [
          MARKET_SEED,
          encodeKey(curator),
          encodeKey(supplyMint),
          encodeKey(collateralMint),
          new Uint8Array([index]),
        ],
        programId,
      );
      return [decodeKey(address), bump];
    },

    findSupplyPositionPda(market, authority) {
      const [address, bump] = findProgramAddress(
        [SUPPLY_POSITION_SEED, encodeKey(market), encodeKey(authority)],
        programId,
      );
      return [decodeKey(address), bump];
    },

    findBorrowPositionPda(market, authority) {
      const [address, bump] = findProgramAddress(
        [BORROW_POSITION_SEED, encodeKey(market), encodeKey(authority)],
        programId,
      );
      return [decodeKey(address), bump];
    },

    findGlobalConfigPda() {
      const [address, bump] = findProgramAddress(
        [GLOBAL_CONFIG_SEED],
        programId,
      );
      return [decodeKey(address), bump];
    },
  };
}
